tools/get_training_info/README.md

```diff
@@ -0,0 +1,303 @@
+* Describe the communication requirements when implementing Data Parallelism, Pipeline Parallelism and Tensor Parallelism. State how many times each dimension needs to communicate, the amount of data transmitted each time and the members of the communication group, among others.
+* Describe the changes in the size of the Transformer model before and after parallel and how these changes impact GPU utilization.
+
+We randomly selected some parallel configurations and used the "Memory Requirement" output in this tool as the predicted value, and the output of "torch.cuda.max_memory_allocated()" in Megatron's [report_memory](../../megatron/utils.py#L81) after training several iterations as the actual value. The parallel configurations in the x-axis of the following figure correspond to the four model parallel configurations in the table below in order.
+
+This can give users insight into whether their planned parallel configuration is trainable, and if it potentially could trigger OOM errors.
+
+<div align="center">
+<img src="https://raw.githubusercontent.com/NVIDIA/Megatron-LM/2ffa75a9bcf6afad563e9be06f6197d8bdb4a814/tools/get_training_info/Mem_Est_vs_Actual.png" alt="图片描述" style="width: 250px; object-fit: cover;">
+</div>
+
+<style>
+table {
+margin: auto;
+}
+</style>
+
+<div  style="zoom:70%" >
+
+Model |Precision | MBS   | GBS       | DP   | PP  | TP  | Peak_Memory_Actual |  Peak_Memory_Estimated | Error (%) 
+:-: |:-: |:-: | :-: | :-:|:-: | :-: | :-: | :-: | :-:
+Llama2 7B | bf16 | 2     |   2048    | 8    | 1   | 1   |  69.1|  68.8 |0.4 
+Llama2 7B | bf16 | 4     |   512     | 4    | 1   | 2   | 55.7 |  55.5 |0.4
+Llama2 7B | bf16 | 2     |   2048    | 4    | 2   | 1   | 49.8|  49.5|0.6
+Llama2 7B | bf16 | 4     |   128     | 1    | 1   | 8   | 28.8 |  28.6 |0.7
+In this table, MBS refers to micro batch size, GBS refers to global batch size, DP denotes data parallelism size, PP denotes pipeline parallelism size, and TP denotes tensor parallelism size.
+
+</div>
+
+
+
+# Calculation Method Explanation
+We analyze the memory requirements of the model parameters, gradients, and optimizer states and the communication behavior of different parallel dimensions based on Megatron([1](https://arxiv.org/pdf/1909.08053.pdf), [2](https://arxiv.org/pdf/2104.04473.pdf), and [3](https://arxiv.org/pdf/2205.05198))
+
+To estimate the memory requirements for the activation portion, given that Megatron supports FlashAttention and Fusion computations, we have adopted a distinctive approach. This method involves collecting the memory address and size information of the corresponding operations each time the cudaMalloc and cudaFree functions are executed, and then conducting line-by-line analysis of this information to derive a computational formula. To implement this method, we used the [torch.cuda.CUDAPluggableAllocator](https://pytorch.org/docs/stable/notes/cuda.html#using-custom-memory-allocators-for-cuda) to customize the memory allocator.
```

Comment:
Thanks for the PR. Could you point out where did you use `torch.cuda.CUDAPluggableAllocator` to estimate the activation memory? I did not find it.

Reply:
@zhipeng93 It is not here. To use it, one must write a shared lib to implement the interface and set it at the beginning the python pytorch program (using ctype to load it).